libvast/src/type.cpp

```diff
@@ -3013,112 +3019,97 @@ record_type::insert_after(std::vector<struct field> fields) noexcept {
-  // the field offset exactly. After having walked over a record type, the
-  // fields are joined back together at the end of the recursive lambda call.
-  const auto do_transform
-    = [](const auto& do_transform, const record_type& self, offset index,
-         auto& current, const auto end) noexcept -> std::optional<record_type> {
-    if (current == end)
-      return self;
-    auto new_fields = std::vector<struct field>{};
-    new_fields.reserve(self.num_fields());
-    index.emplace_back(self.num_fields());
-    while (index.back() > 0 && current != end) {
-      const auto& old_field = self.field(--index.back());
-      // Compare the offsets of the next target with our current offset.
-      const auto [index_mismatch, current_index_mismatch]
-        = std::mismatch(index.begin(), index.end(), current->index.begin(),
-                        current->index.end());
-      if (index_mismatch == index.end()
-          && current_index_mismatch == current->index.end()) {
-        // The offset matches exactly, so we apply the transformation.
-        do {
-          auto replacements = std::invoke(std::move(current->fun), old_field);
-          std::move(replacements.rbegin(), replacements.rend(),
-                    std::back_inserter(new_fields));
-          ++current;
-        } while (current != end && current->index == index);
-      } else if (index_mismatch == index.end()) {
-        // The index is a prefix of the target offset for the next
-        // transformation, so we recurse one level deeper.
-        VAST_ASSERT(caf::holds_alternative<record_type>(old_field.type));
-        if (auto sub_result
-            = do_transform(do_transform, caf::get<record_type>(old_field.type),
-                           index, current, end))
-          new_fields.push_back({
-            std::string{old_field.name},
-            std::move(*sub_result),
-          });
-        // Check for invalid arguments on the way in.
-        VAST_ASSERT(current == end || index != current->index,
-                    "cannot apply transformations to both a nested record type "
-                    "and its children at the same time.");
+  VAST_ASSERT(std::is_sorted(transformations.begin(), transformations.end()),
+              "transformations must be sorted by index");
+  VAST_ASSERT(transformations.end()
+                == std::adjacent_find(
+                  transformations.begin(), transformations.end(),
+                  [](const auto& lhs, const auto& rhs) noexcept {
+                    const auto [lhs_mismatch, rhs_mismatch]
+                      = std::mismatch(lhs.index.begin(), lhs.index.end(),
+                                      rhs.index.begin(), rhs.index.end());
+                    return lhs_mismatch == lhs.index.end();
+                  }),
+              "transformation indices must not be a subset of the following "
+              "transformation's index");
```

Comment:
> Would it make sense to split this and get just the bugfix in?

This actually fixed another bug for record types like this:

```
type foo = record {
  a: record {
    b: record {
      c: count,
    },
  },
  d: count,
}
```

Where before this change we returned the invalid type

```
type foo = record {
  a: record {
  },
  d: count,
}
```

rather than the correct:

```
type foo = record {
  d: count,
}
```

So no, I think we actually need to get that fix in as well. Users were theoretically able to trigger it with carefully crafted data.

Regarding `VAST_ASSERT_EXPENSIVE`: Yes, I'm all for that. We can enable that in Debug builds only and leave it disabled in RelWithDebInfo.

Reply:
I think in this case we could also argue that the function should return an error rather than assert, because I don't think the check is all too expensive. We can probably fix that after the release, though. Compared to the actual transformation and reassembly of potentially nested record types this check should be cheap.

Reply:
I created sc-33355 to track this.